### PR TITLE
Add delete binding action to provisioned services

### PR DESCRIPTION
**Pull request by benjaminapetersen**
https://trello.com/c/JMKxStDt

- This is my `unbind-service` branch w/spadgett's `Update delete binding from service-instance-row to use app names` cherry-pick.  
- Rebased to eliminate the cherry-pick commit

I may close the previous `unbind-service` PR & keep this on to avoid the need to fiddle around with merge conflicts again.

Prev #1505

### CODE REVIEW

**Comment by benjaminapetersen** [comment]
From a service instance, bindings to apps list `kind` & `app name`, otherwise if a secret lists the binding name:

screen shot 2017-06-13 at 11 50 00 pm

Bindings from apps list the binding name:
screen shot 2017-06-13 at 11 51 29 pm

**Comment by benjaminapetersen** [comment]
@jwforres @spadgett pretty sure I'll close the other one & squash this, just thinking on it a bit.

File: app/scripts/services/serviceInstances.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+angular.module('openshiftConsole')
+  .factory('ServiceInstancesService', function() {
+    return {
+      // NOTE: should this be a filter?  or will shared functionality grow here?
```
Comment (spadgett) [inline app/scripts/services/serviceInstances.js]: I think this should be a filter. Maybe more than one: `serviceClassDisplayName` and `serviceInstanceDisplayName`?

File: app/scripts/services/bindingModalUtils.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+'use strict';
+
+angular.module('openshiftConsole')
+  .factory('BindingModalUtils', [
```
Comment (spadgett) [inline app/scripts/services/bindingModalUtils.js]: Instead of making it specific to the binding modal, this could be a general utility that combines and (optionally?) sorts collections of API objects.

We have this need for config maps and secrets in some places where we combine them in a select.

File: app/scripts/services/bindingModalUtils.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+                                    deployments,
+                                    replicationControllers,
+                                    replicaSets,
+                                    statefulSets) {
+          // Don't waste time sorting on each data load, just sort when we have them all
+          if (deploymentConfigs && deployments && replicationControllers && replicaSets && statefulSets) {
+            var apiObjects = deploymentConfigs.concat(deployments)
+                                              .concat(replicationControllers)
+                                              .concat(replicaSets)
+                                              .concat(statefulSets);
```
Comment (spadgett) [inline app/scripts/services/bindingModalUtils.js]: Probably want to check if these are arrays before calling `concat`. Or at least make a comment that the inputs must be arrays.

File: app/scripts/directives/bindService.js [hunk 4]
```diff
@@ -55,20 +59,6 @@
...
-    var deploymentConfigs, deployments, replicationControllers, replicaSets, statefulSets;
-    var sortApplications = function() {
-      // Don't waste time sorting on each data load, just sort when we have them all
-      if (deploymentConfigs && deployments && replicationControllers && replicaSets && statefulSets) {
-        var apiObjects = deploymentConfigs.concat(deployments)
-                                          .concat(replicationControllers)
-                                          .concat(replicaSets)
-                                          .concat(statefulSets);
-        ctrl.applications = _.sortByAll(apiObjects, ['metadata.name', 'kind']);
-        ctrl.bindType = ctrl.applications.length ? "application" : "secret-only";
```
Comment (spadgett) [inline app/scripts/directives/bindService.js]: You are no longer setting `ctrl.bindType`

File: app/scripts/directives/bindService.js [hunk 5]
```diff
@@ -55,20 +59,6 @@
       }
     };
 
-
-    var deploymentConfigs, deployments, replicationControllers, replicaSets, statefulSets;
-    var sortApplications = function() {
```
Comment (spadgett) [inline app/scripts/directives/bindService.js]: Suggest leaving `sortApplications` as a function then calling your new service function from here so you don't have to repeat the parameters in every list callback.

```js
var sortApplication = function() {
  ctrl.applications = BindingModalUtils.sortApplications(deploymentConfigs, deployments, replicationControllers, replicaSets, statefulSets);
  ctrl.bindType = _.isEmpty(ctrl.applications) ? "secret-only" : "application";
};
```

File: app/scripts/directives/unbindService.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+  function UnbindService($filter, BindingModalUtils, DataService, ServiceInstancesService) {
+    var ctrl = this;
+    var context;
+    var getDisplayName = ServiceInstancesService.getDisplayName;
+    // var sortApplications = BindingModalUtils.sortApplications;
+    // TODO: once we have podPreset
+    // var deploymentConfigs, deployments, replicationControllers, replicaSets, statefulSets;
+
+
+    // TODO: once we have podPreset we will be able to associate the
```
Comment (spadgett) [inline app/scripts/directives/unbindService.js]: Remove the TODO comments?

File: app/scripts/directives/unbindService.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+      'BindingModalUtils',
+      'DataService',
+      'ServiceInstancesService',
+      UnbindService
+    ],
+    controllerAs: 'ctrl',
+    bindings: {
+      target: '<',
+      bindings: '<',
+      applications: '<',
```
Comment (spadgett) [inline app/scripts/directives/unbindService.js]: `bindings` and `applications` are maps. Looking at just the names, I thought these were flat lists. Can we use similar names to what we use on the overview so it's clear they are the same maps?

File: app/scripts/directives/unbindService.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        onShow: showConfirm
+      }];
+
+      context = {
+        namespace: _.get(ctrl.target, 'metadata.namespace')
+      };
+    };
+
+    ctrl.appNameForBinding = function(bindingName) {
+      return _.get(_.first(_.get(ctrl.applications, bindingName)), ['metadata', 'name']) || bindingName;
```
Comment (spadgett) [inline app/scripts/directives/unbindService.js]: I find the nested lodash calls on one line here and just below hard to follow.

Comment (spadgett) [inline app/scripts/directives/unbindService.js]: It looks like we're showing just the first app name? Usually there is only one app, but if unbind is going to affect multiple apps, it's important to know.

File: app/views/overview/_list-row-actions.html [hunk 9]
```diff
@@ -28,6 +28,9 @@
             <li ng-if="('pod_presets' | enableTechPreviewFeature) && row.state.bindableServiceInstances.length" role="menuitem">
               <a href="" ng-click="row.showOverlayPanel('bindService', {target: row.apiObject})">Create Binding</a>
             </li>
+            <li ng-if="('pod_presets' | enableTechPreviewFeature) && row.state.bindableServiceInstances.length" role="menuitem">
```
Comment (spadgett) [inline app/views/overview/_list-row-actions.html]: Maybe use `ng-if-start` on the `li` above and `ng-if-end` since the conditions are the same?

File: app/scripts/services/bindingModalUtils.js (hunk 2 above)
Comment (spadgett) [inline app/scripts/services/bindingModalUtils.js]: Is this only used in the bind modal and no longer needed for unbind? I suggest for now we leave the working bind code alone if we're not reusing it for unbind.

**Comment by benjaminapetersen** [comment]
@spadgett yup! bring on the reviews.  Not exactly ready like I'd like to be, but time is of the essence.  I'll hit these items while I work on cleanup.

File: app/scripts/services/serviceInstances.js (hunk 1 above)
Comment (benjaminapetersen) [inline app/scripts/services/serviceInstances.js]: Agree.  This will become `serviceInstanceDisplayName`.  

For a `serviceClassDisplayName` filter, are you thinking we need it because the `externalMetadata` field is a different than the typical `metadata.name`?

Comment (benjaminapetersen) [inline app/scripts/services/serviceInstances.js]: Moved the function to a filter, left a comment for a `TODO` on `serviceClassDisplayName` per my previous comment.

File: app/scripts/services/bindingModalUtils.js (hunk 2 above)
Comment (benjaminapetersen) [inline app/scripts/services/bindingModalUtils.js]: Yeah, probably a premature refactor on my part, assuming I would need it for `unbind`. I  will return this to its previous state.

File: app/scripts/services/bindingModalUtils.js (hunk 3 above)
Comment (benjaminapetersen) [inline app/scripts/services/bindingModalUtils.js]: agree, I'll start with `[].concat(deploymentConfigs)` to ensure we don't error here.

File: app/scripts/directives/bindService.js (hunk 5 above)
Comment (benjaminapetersen) [inline app/scripts/directives/bindService.js]: I moved it back to `bindService` as it originally was. Def think this is a good call if we end up sharing it in the future.

File: app/scripts/directives/bindService.js (hunk 4 above)
Comment (benjaminapetersen) [inline app/scripts/directives/bindService.js]: fixed

File: app/scripts/directives/unbindService.js (hunk 7 above)
Comment (benjaminapetersen) [inline app/scripts/directives/unbindService.js]: `bindingsByInstanceRef` && `applicationsByBinding`, will do.

File: app/scripts/services/bindingModalUtils.js (hunk 3 above)
Comment (spadgett) [inline app/scripts/services/bindingModalUtils.js]: I think it's more than this... If you pass an object as an argument to concat, does it just append the object?

Maybe a moot point if you're rolling back the retractor though

Comment (benjaminapetersen) [inline app/scripts/services/bindingModalUtils.js]: Yeah... everything below has a `_.toArray` wrap (original code before my change).  If we do move it out in future, agree, perhaps the `_.toArray` should come with it & happen internally.

File: app/scripts/directives/unbindService.js (hunk 6 above)
Comment (benjaminapetersen) [inline app/scripts/directives/unbindService.js]: Eliminated this whole section, don't think any of it will ever be needed w/the maps from overview.

File: app/scripts/directives/unbindService.js (hunk 8 above)
Comment (benjaminapetersen) [inline app/scripts/directives/unbindService.js]: Updating to this:

screen shot 2017-06-14 at 1 19 32 pm

@ncameronbritt for input if this deviates or is new.  In theory, if a binding represents multiple apps, this is a bit wasteful of space as they could be listed horizontally, but I don't know of another pattern off the top of my head so going with consistency for now.  We can iterate if needed.

File: app/views/overview/_list-row-actions.html (hunk 9 above)
Comment (benjaminapetersen) [inline app/views/overview/_list-row-actions.html]: will do

**Comment by spadgett** [comment]
For the radio buttons, wonder if it should be 

```html
{{name}} <small class="text-muted">&ndash; -- {{kind}}</small>`
```

like in the bind application dropdown?

openshift web console 2017-06-14 13-41-19

**Comment by benjaminapetersen** [comment]
@spadgett @jwforres finished round 1 of requested changes. I'm looking it over again myself, but feel free to give it another pass.  Def prefer a goal of pre 3pm to get it in the queue :)

**Comment by benjaminapetersen** [comment]
@spadgett yeah good Q.  We have all of these:
screen shot 2017-06-14 at 1 45 39 pm
screen shot 2017-06-14 at 1 45 43 pm
screen shot 2017-06-14 at 1 45 47 pm

I like your suggestion, I think it would improve density / clarity.

File: app/views/directives/bind-service/remove-binding-confirm.html [hunk 10]
```diff
@@ -0,0 +1,21 @@
+<div>
+  <div ng-if="!ctrl.error">
+    <h3 class="mar-top-none">
+      Binding for {{ctrl.appKindForBinding(ctrl.selectedBinding) | humanizeKind : true}} <strong>{{ctrl.appNameForBinding(ctrl.selectedBinding)}}</strong> has been deleted.
```
Comment (spadgett) [inline app/views/directives/bind-service/remove-binding-confirm.html]: Looks like you're still calling a method you removed, `appNameForBinding`?

**Comment by spadgett** [comment]
On the last binding dialog, we show the binding name, but we don't show the name anywhere on the overview currently, so I'm not going to know which to delete.

File: app/views/directives/bind-service/remove-binding-confirm.html (hunk 10 above)
Comment (benjaminapetersen) [inline app/views/directives/bind-service/remove-binding-confirm.html]: Fixing that now, working on listing here as well.

**Comment by spadgett** [comment]
I think the first wizard step label should either be "Services" or "Applications" depending on what we're showing to be consistent with the Create Binding dialog.

**Comment by benjaminapetersen** [comment]
Density here definitely improves clarity:

screen shot 2017-06-14 at 2 04 41 pm

File: app/views/directives/bind-service/select-binding.html [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+            <div class="component-label">
+              {{ (appForBinding.kind) | humanizeKind : true}}
+            </div>
+            <p>{{appForBinding.metadata.name || binding.metadata.name}}</p>
+          </div>
+          <div ng-if="!(ctrl.appsForBinding(binding))">
+            <div class="component-label">
+              Secret
+            </div>
+            <p>{{appForBinding.metadata.name || binding.metadata.name}}</p>
```
Comment (spadgett) [inline app/views/directives/bind-service/select-binding.html]: Should this be...

```html
<p>{{binding.spec.secretName}}</p>
```

**Comment by spadgett** [comment]
Need to disable the Delete button when nothing is selected or preselect the first item. I think preselecting the first is OK? @ncameronbritt @jwforres 

openshift web console 2017-06-14 14-17-05

**Comment by ncameronbritt** [comment]
I think preselecting if there is only one makes sense. I think if there is more than one we wouldn't preselect, and the "Delete" button would be disabled. I believe that is consistent with the Bind flow.

**Comment by benjaminapetersen** [comment]
Updating confirmation to list apps & match first step in panel:

screen shot 2017-06-14 at 2 22 32 pm

Thoughts?  (I see I have more comments to loop back on, have not addressed yet)

File: app/views/directives/bind-service/select-binding.html (hunk 11 above)
Comment (benjaminapetersen) [inline app/views/directives/bind-service/select-binding.html]: Yup will do

**Comment by benjaminapetersen** [comment]
@spadgett @ncameronbritt  if first should be "Applications" or "Services", should last step also be "Results"?  The word "Confirmation" implies to me I would have one more changes to back out of the action.

**Comment by spadgett** [comment]
Yes, should be "Results". I believe that's what we use everywhere else.

**Comment by benjaminapetersen** [comment]
@spadgett @ncameronbritt working on the pre-select / disable now.  I think that requires this `validityWatcher` like bind service.  FWIW, I'd like to look at updating `pf-wizard` to support this is a more standard way, its a little kludgy, may open an issue since validation is pretty common.

File: app/index.html [hunk 12]
```diff
@@ -337,6 +337,7 @@ <h2 class="text-center" id="temporary-loading-message" style="display: none;">Lo
         <script src="scripts/directives/buildStatus.js"></script>
         <script src="scripts/directives/routeServiceBarChart.js"></script>
         <script src="scripts/directives/bindService.js"></script>
+        <script src="scripts/directives/unbindService.js"></script>
```
Comment (benjaminapetersen) [inline app/index.html]: I prob should rename this file since we no longer call it "unbind"

File: app/views/overview/_list-row.html [hunk 13]
```diff
@@ -18,4 +18,10 @@
     <bind-service target="row.overlay.state.target"
                   on-close="row.closeOverlayPanel"></bind-service>
   </div>
+  <div ng-if="row.overlay.panelName === 'unbindService'">
+    <unbind-service
+      target="row.overlay.state.target"
+      bindings-by-instance="row.bindings"
```
Comment (spadgett) [inline app/views/overview/_list-row.html]: Ah, so it's not a map. I'd call the attribute bindings if it's just the array of bindings. Sorry if I misunderstood.

File: app/views/directives/bind-service/delete-binding-result.html [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+      Binding for the following has been deleted:
+    </h3>
+
+    <div
+      ng-if="ctrl.appsForBinding(ctrl.selectedBinding)"
+      ng-repeat="appForBinding in ctrl.appsForBinding(ctrl.selectedBinding)">
+      {{appForBinding.metadata.name}} <small class="text-muted">&ndash; {{ appForBinding.kind | humanizeKind : true}}</small>
+    </div>
+    <div ng-if="!(ctrl.appsForBinding(ctrl.selectedBinding))">
+      {{ctrl.selectedBinding}} <small class="text-muted">&ndash; Secret</small>
```
Comment (spadgett) [inline app/views/directives/bind-service/delete-binding-result.html]: This is always the binding name here and not a secret correct?

Comment (spadgett) [inline app/views/directives/bind-service/delete-binding-result.html]: I think you can put `Binding` here for now and we can circle back. I have some concerns that we don't show the binding name anywhere in the UI, but we can fix as a follow on to get the basic functionality in.

Comment (benjaminapetersen) [inline app/views/directives/bind-service/delete-binding-result.html]: I'll swap this & leave a `TODO:` because its weird to select "secret" then see "binding" deleted, agree. :/

File: app/views/overview/_list-row.html (hunk 13 above)
Comment (benjaminapetersen) [inline app/views/overview/_list-row.html]: Ok, here will be: 
```html
  <unbind-service
      target="row.overlay.state.target"
      binding="row.bindings"
      on-close="row.closeOverlayPanel"></unbind-service>
```
and in the service instance row template:
```html
    <unbind-service
      target="row.overlay.state.target"
      bindings="row.bindings"
      applications-by-binding="row.state.applicationsByBinding"
      on-close="row.closeOverlayPanel"></unbind-service>
```
So we will have `bindings` and `applicationsByBinding`.

**Comment by spadgett** [comment]
Thanks for the fast turnaround @benjaminapetersen 

[merge]

**Comment by benjaminapetersen** [comment]
woohoo! :yum:
